src/Formik.tsx

```diff
@@ -441,16 +443,39 @@ export class Formik<Values = object, ExtraProps = {}> extends React.Component<
       submitCount: prevState.submitCount + 1,
     }));
 
-    return this.runValidations(this.state.values).then(combinedErrors => {
-      this.setState({ isValidating: false });
-      const isValid = Object.keys(combinedErrors).length === 0;
-      if (isValid) {
-        this.executeSubmit();
-      } else if (this.didMount) {
-        // ^^^ Make sure Formik is still mounted before calling setState
-        this.setState({ isSubmitting: false });
+    return this.runValidations(this.state.values).then(
+      combinedErrors => {
+        this.setState({ isValidating: false });
+        // In case an error was thrown and passed to the resolved Promise,
+        // `combinedErrors` can be an instance of an Error. We need to check
+        // that and abort the submit.
+        // If we don't do that, calling `Object.keys(new Error())` yields an
+        // empty array, which causes the validation to pass and the form
+        // to be submitted.
+        const isInstanceOfError = combinedErrors instanceof Error;
+        const isValid =
+          !isInstanceOfError && Object.keys(combinedErrors).length === 0;
+        if (isValid) {
+          this.executeSubmit();
+        } else if (this.didMount) {
+          // ^^^ Make sure Formik is still mounted before calling setState
+          this.setState({ isSubmitting: false });
+        }
+      },
+      error => {
```

Comment:
I'm not sure if this is really necessary but I felt like we could gracefully handle this and show a warning.

If you think it's not necessary, I can remove it.